core/src/main/java/io/grpc/internal/DelayedClientTransport.java

```diff
@@ -139,33 +124,30 @@ public final ClientStream newStream(
     try {
       PickSubchannelArgs args = new PickSubchannelArgsImpl(
           method, headers, callOptions, new PickDetailsConsumerImpl(tracers));
-      SubchannelPicker picker = null;
-      long pickerVersion = -1;
+      PickerState state = pickerState;
       while (true) {
-        synchronized (lock) {
-          if (shutdownStatus != null) {
-            return new FailingClientStream(shutdownStatus, tracers);
-          }
-          if (lastPicker == null) {
-            return createPendingStream(args, tracers);
-          }
-          // Check for second time through the loop, and whether anything changed
-          if (picker != null && pickerVersion == lastPickerVersion) {
-            return createPendingStream(args, tracers);
-          }
-          picker = lastPicker;
-          pickerVersion = lastPickerVersion;
+        if (state.shutdownStatus != null) {
+          return new FailingClientStream(state.shutdownStatus, tracers);
         }
-        PickResult pickResult = picker.pickSubchannel(args);
-        ClientTransport transport = GrpcUtil.getTransportFromPickResult(pickResult,
-            callOptions.isWaitForReady());
-        if (transport != null) {
-          return transport.newStream(
-              args.getMethodDescriptor(), args.getHeaders(), args.getCallOptions(),
-              tracers);
+        if (state.lastPicker != null) {
+          PickResult pickResult = state.lastPicker.pickSubchannel(args);
+          ClientTransport transport = GrpcUtil.getTransportFromPickResult(pickResult,
+              callOptions.isWaitForReady());
+          if (transport != null) {
+            return transport.newStream(
+                args.getMethodDescriptor(), args.getHeaders(), args.getCallOptions(),
+                tracers);
+          }
         }
         // This picker's conclusion is "buffer".  If there hasn't been a newer picker set (possible
```

Comment:
This comment doesn't make sense, particularly since there is no variable named "buffer".

Reply:
"buffer" == "buffer the RPC" == `PickResult.withNoResult()` in this context. I wouldn't mind the comment being more clear (either change "it" or explain "buffer the RPC"). But it isn't directly related to the PR, so I'll let that be a follow-up.

Reply:
https://github.com/grpc/grpc-java/pull/11242